header-rewrite-filter/header_rewrite.cc

```diff
@@ -118,9 +135,13 @@ Http::FilterHeadersStatus HttpHeaderRewriteFilter::encodeHeaders(Http::ResponseH
   // execute each operation
   for (auto const& processor : response_header_processors_) {
     processor->executeOperation(headers);
-    ENVOY_LOG_MISC(info, "added response header!"); // TODO: remove debug statement once response-side test setup is created
   }
 
+  // TODO: remove debug statement
+  // const bool result = set_bool_processors_.at("mock_bool")->executeOperation();
```

Comment:
I'm not sure this is going to work. I think all the operations will need to execute in order, because some might change things that affect others. Which also means the container needs to be ordered.

I think the reason the polymorphism doesn't work is also because `BoolProcessor` is trying to only do the things it needs to, i.e. not take headers. But in this case it's necessary to be able to "plug in" any processor.

Rather, treat it like an Event system. The Envoy functions `decodeHeaders`, `decodeData`, `encodeHeaders`, `encodeData` are "hooks". And each Processor type can execute on whatever hooks it cares about, or not.

Suggestion:
1. Instead of just `execute`, give Processor every hook: `decodeHeaders`, `decodeData`, `encodeHeaders`, `encodeData`
    - this will also fix the issue that the code currently doesn't distinguish between Requests and Responses. We need some way to say "set these headers in the upstream requests" vs "set these headers in the response to the client".  HAProxy does this in the directive, `http-response add-header` vs `http-request add-header`; ATS does it with hooks, `cond %{SEND_RESPONSE_HDR_HOOK} set-header` vs cond %{SEND_REQUEST_HDR_HOOK} set-header`. What I'm suggesting is to make your _implementation_ use hooks, just to make it easier; but your config language can look like whatever you want.
3. Don't make them Pure Virtual. Give them empty functions in Processor: 
`Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap&, bool) override { return Http::FilterHeadersStatus::Continue; }`
4. Then, each Processor subclass can implement the hooks it cares about, and the ones it won't will get the default do-nothing of the base class.
5. Change `response_header_processors_` and `set_bool_processors_` to be a single `vector<unique_ptr<Processor>>`
6. Then, `HttpHeaderRewriteFilter::decodeHeaders` will iterate over the vector and call `decodeHeaders` on each Processor (likewise for the other hooks).
    - this also gives us the power in the future to let a directive/Processor stop execution of later directives, if we want. `Processor::decodeHeaders` doesn't have to return a `FilterHeadersStatus`, it could return a bool that means "continue with the next directive, or halt". Or whatever we want.
 7. `BoolProcessor` will implement `Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap&, bool)`. It doesn't actually need the headers, but that lets it share the function with the base class, so `processor->decodeHeaders` on the base class from the `vector<Procesor>` works

Does that make sense? I think that will make it easier to work with, and solve a lot of the polymorphism issues. But it's just a suggestion, feel free to do whatever works

Reply:
I see what you're saying about each Processor calling encodeHeaders! I'm not sure if these hooks can be called at will by the processors because they are well-defined functions in the broader Envoy event loop. Will look into this though.

> this will also fix the issue that the code currently doesn't distinguish between Requests and Responses.

I think the current implementation does handle performing an operation on only the request side or only the response side. (The first keyword in each operation `http-request` vs `http-response` handles this).

> this also gives us the power in the future to let a directive/Processor stop execution of later directives, if we want. Processor::decodeHeaders doesn't have to return a FilterHeadersStatus, it could return a bool that means "continue with the next directive, or halt".

The current structure returns back to `encodeHeaders` or `decodeHeaders` after each processor is done executing, so each processor could return an instruction to the filter to keep executing more operations or halt.

> Change response_header_processors_ and set_bool_processors_ to be a single vector<unique_ptr<Processor>>

This was my first approach when I was building the `SetBoolProcessor`, but I couldn't find a good way to cast a `Processor` to a `HeaderProcessor` -- is there a way to cast a pointer of an abstract base class to a pointer of a child class that is also abstract?

Reply:
> This was my first approach when I was building the SetBoolProcessor, but I couldn't find a good way to cast a Processor to a HeaderProcessor -- is there a way to cast a pointer of an abstract base class to a pointer of a child class that is also abstract?

I was suggesting to avoid that. IMO casting is generally a sign that the inheritance structure has issues.

You can use `static_cast`:
```
#include <iostream>
#include <string>
#include <memory>

class BaseAbstract {
public:
  virtual ~BaseAbstract() {}
  virtual int somef() = 0;   
};

class ChildAbstract : public BaseAbstract {
public:
  virtual ~ChildAbstract() {}
  virtual int somef() = 0; 
  int customf() { return 9; }
};

class ChildConcrete : public ChildAbstract {
public:
  virtual ~ChildConcrete() {}
  virtual int somef() { return 42; }   
};


int main()
{
  std::unique_ptr<BaseAbstract> basep(new ChildConcrete());
  BaseAbstract* base_raw_ptr = basep.get();
  ChildAbstract* childap = static_cast<ChildAbstract*>(base_raw_ptr);
  std::cout << childap->customf() << "\n";
}
```

[Playground](https://cpp.sh/?source=%23include+%3Ciostream%3E%0A%23include+%3Cstring%3E%0A%23include+%3Cmemory%3E%0A%0Aclass+BaseAbstract+%7B%0Apublic%3A%0A++virtual+~BaseAbstract()+%7B%7D%0A++virtual+int+somef()+%3D+0%3B+++%0A%7D%3B%0A%0Aclass+ChildAbstract+%3A+public+BaseAbstract+%7B%0Apublic%3A%0A++virtual+~ChildAbstract()+%7B%7D%0A++virtual+int+somef()+%3D+0%3B+%0A++int+customf()+%7B+return+9%3B+%7D%0A%7D%3B%0A%0Aclass+ChildConcrete+%3A+public+ChildAbstract+%7B%0Apublic%3A%0A++virtual+~ChildConcrete()+%7B%7D%0A++virtual+int+somef()+%7B+return+42%3B+%7D+++%0A%7D%3B%0A%0A%0Aint+main()%0A%7B%0A++std%3A%3Aunique_ptr%3CBaseAbstract%3E+basep(new+ChildConcrete())%3B%0A++BaseAbstract*+base_raw_ptr+%3D+basep.get()%3B%0A++ChildAbstract*+childap+%3D+static_cast%3CChildAbstract*%3E(base_raw_ptr)%3B%0A++std%3A%3Acout+%3C%3C+childap-%3Ecustomf()+%3C%3C+%22%5Cn%22%3B%0A%7D)

But the cast can fail at runtime if there's a bug or the control flow isn't what's expected.

That was the idea behind my suggestion: if you can make the base class share the function, you can avoid needing to cast, and all the dangers and potential bugs that come with it

Reply:
Oh I see! That will be a much nicer implementation, thanks for clarifying :))

Reply:
> I think all the operations will need to execute in order, because some might change things that affect others. Which also means the container needs to be ordered.

agreed

Reply:
Modified the interface to not have any pure virtual functions. I'm still confused about the ordering discussion though -- `encodeHeaders` and `decodeHeaders` will execute operations in the same order as in the config. So isn't it trivial to stop executing the rest of the operations if anything goes wrong, or in other words we can assume the ordering is already present in the filter sequence?